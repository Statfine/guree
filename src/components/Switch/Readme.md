### 基本用法

```jsx
<Switch onChange={v => console.log(v)}></Switch>
```

### 设置选中/未选中显示的内容

```jsx
<Switch checkedChildren={"checked"} unCheckedChildren={"unchecked"}></Switch>
```

### 默认选中
```jsx
<Switch defaultChecked></Switch>
```

### 禁用状态

```jsx
<Switch disabled></Switch>
```

### 受控组件

```jsx
import Button from "@components/Button";
import { useState } from "react";
const [checked, setChecked] = useState(false)
const change = (value) => {
  setChecked(!checked)
}

<div>
  <Switch defaultChecked checked={checked}></Switch>
  <br />
  <br />
  <Button onClick={change}>switch</Button>
</div>

```
