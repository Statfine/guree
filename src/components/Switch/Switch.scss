@import "../../common/utils/index.scss";

$switch: scopedClass(switch);

.#{$switch} {
  min-width: 44px;
  min-height: 22px;
  position: relative;
  padding: 2px;
  border-radius: 11px;
  background-color: rgba(0, 0, 0, 0.25);
  color: #fff;
  display: inline-flex;
  cursor: pointer;
  &-handler {
    width: 18px;
    height: 18px;
    position: absolute;
    background-color: #fff;
    left: 2px;
    border-radius: 50%;
    top: 2px;
    transition: all 0.36s cubic-bezier(0.78, 0.14, 0.15, 0.86);
  }
  &-inner {
    min-height: 18px;
    min-width: 16px;
    @include flex-center-center;
    position: absolute;
    right: 7px;
    top: 2px;
    transition: all 0.36s cubic-bezier(0.78, 0.14, 0.15, 0.86);
  }

  &-checked {
    background-color: #409eff;
    .#{$switch}-handler {
      left: calc(100% - 18px - 2px);
    }
    .#{$switch}-inner {
      right: 25px;
    }
  }
  &-unvisible {
    visibility: hidden;
    display: inline-flex;
    .#{$switch}-handler {
      position: unset;
    }
    .#{$switch}-inner {
      position: unset;
      margin-left: 5px;
      margin-right: 5px;
    }
  }

  &-disabled {
    cursor: not-allowed;
    opacity: 0.4;
    &.#{$switch}-checked {
      background-color: rgba(0, 0, 0, 0.25);
    }
  }
}
