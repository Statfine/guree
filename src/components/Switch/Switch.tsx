import React, { ReactNode, MouseEventHandler, useState } from "react";
import "./Switch.scss";
import {
  combineClasses,
  scopedClass,
  classesObj,
  updateStateOnPropChange,
} from "@utils/index";
import { isUndefined } from "util";

interface Props {
  /** 选中状态 */
  value?: boolean;
  /** 默认选中状态 */
  defaultValue?: boolean;
  /** 选中时的内容 */
  checkedChildren?: ReactNode;
  /** 未选中时的内容 */
  unCheckedChildren?: ReactNode;
  /** 禁用状态 */
  disabled?: boolean;
  /** 状态改变时的回调 */
  onChange?: (checked: boolean) => {};
}

const Switch: React.FunctionComponent<Props> = ({
  value,
  defaultValue,
  checkedChildren,
  unCheckedChildren,
  disabled = false,
  onChange,
}: Props) => {
  const switchClass = (...classes: (string | Array<string> | classesObj)[]) =>
    scopedClass("switch", ...classes);

  const [checkedStatus, setCheckedStatus] = useState(
    isUndefined(value) ? defaultValue : value
  );

  const switchClassName = combineClasses({
    [switchClass()]: true,
    [switchClass("checked")]: Boolean(checkedStatus),
    [switchClass("disabled")]: disabled,
  });

  const change: MouseEventHandler<HTMLElement> = (e) => {
    onChange && onChange(!checkedStatus);
    if (disabled || !isUndefined(value)) {
      return;
    }
    setCheckedStatus(!checkedStatus);
  };
  updateStateOnPropChange(value, setCheckedStatus);

  return (
    <span className={switchClassName} onClick={change}>
      <span className={switchClass("unvisible")}>
        <span className={switchClass("handler")}></span>
        <span className={switchClass("inner")}>
          {checkedStatus ? checkedChildren : unCheckedChildren}
        </span>
      </span>
      <span className={switchClass("handler")}></span>
      <span className={switchClass("inner")}>
        {checkedStatus ? checkedChildren : unCheckedChildren}
      </span>
    </span>
  );
};

export default Switch;
