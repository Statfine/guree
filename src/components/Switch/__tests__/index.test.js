import React from "react";
import renderer from "react-test-renderer";
import Switch from "..";
import { mount } from "enzyme";

describe("Switch 组件", () => {
  it("正确导出 Switch 组件", () => {
    const sw = renderer.create(<Switch></Switch>).toJSON();
    expect(sw).toMatchSnapshot();
  });

  it("测试 Switch 组件 checkedChildren 参数", () => {
    const sw = renderer.create(<Switch checkedChildren={"checked"}></Switch>).toJSON();
    expect(sw).toMatchSnapshot();
  });

  it("测试 Switch 组件 unCheckedChildren 参数", () => {
    const sw = renderer.create(<Switch unCheckedChildren={"unchecked"}></Switch>).toJSON();
    expect(sw).toMatchSnapshot();
  });

  it("测试 Switch 组件 defaultChecked 参数", () => {
    const sw = renderer.create(<Switch defaultChecked></Switch>).toJSON();
    expect(sw).toMatchSnapshot();
  });

  it("测试 Switch 组件 disabled 参数", () => {
    const sw = renderer.create(<Switch disabled></Switch>).toJSON();
    expect(sw).toMatchSnapshot();
  });

  it("测试 Switch 组件 checked 参数", () => {
    const sw = renderer.create(<Switch checked></Switch>).toJSON();
    expect(sw).toMatchSnapshot();
  });

  it("测试 Switch 组件 onChange 参数", () => {
    const fn = jest.fn()
    const sw = mount(<Switch onChange={fn}></Switch>);
    sw.simulate("click")
    expect(fn).toBeCalledWith(true)
    sw.update()
    expect(sw.find(".guree-switch").hasClass("guree-switch-checked")).toEqual(true);
    sw.simulate("click")
    expect(fn).toBeCalledWith(false)
    sw.update()
    expect(sw.find(".guree-switch").hasClass("guree-switch-checked")).toEqual(false);
  });

  it("测试 Switch 组件 disabled 状态下的点击事件", () => {
    const sw = mount(<Switch disabled></Switch>);
    sw.simulate("click")
    sw.update()
    expect(sw.find(".guree-switch").hasClass("guree-switch-checked")).toEqual(false);
  });

  it("测试 Switch 组件 checked 不为 undefined 状态下的点击事件", () => {
    const sw1 = mount(<Switch checked={true}></Switch>);
    const sw2 = mount(<Switch checked={false}></Switch>);
    sw1.simulate("click")
    sw2.simulate("click")
    sw1.update()
    sw2.update()
    expect(sw1.find(".guree-switch").hasClass("guree-switch-checked")).toEqual(true);
    expect(sw2.find(".guree-switch").hasClass("guree-switch-checked")).toEqual(false);
  });
});
