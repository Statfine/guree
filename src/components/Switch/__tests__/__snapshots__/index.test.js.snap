// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch 组件 正确导出 Switch 组件 1`] = `
<span
  className="guree-switch"
  onClick={[Function]}
>
  <span
    className="guree-switch-unvisible"
  >
    <span
      className="guree-switch-handler"
    />
    <span
      className="guree-switch-inner"
    />
  </span>
  <span
    className="guree-switch-handler"
  />
  <span
    className="guree-switch-inner"
  />
</span>
`;

exports[`Switch 组件 测试 Switch 组件 checked 参数 1`] = `
<span
  className="guree-switch guree-switch-checked"
  onClick={[Function]}
>
  <span
    className="guree-switch-unvisible"
  >
    <span
      className="guree-switch-handler"
    />
    <span
      className="guree-switch-inner"
    />
  </span>
  <span
    className="guree-switch-handler"
  />
  <span
    className="guree-switch-inner"
  />
</span>
`;

exports[`Switch 组件 测试 Switch 组件 checkedChildren 参数 1`] = `
<span
  className="guree-switch"
  onClick={[Function]}
>
  <span
    className="guree-switch-unvisible"
  >
    <span
      className="guree-switch-handler"
    />
    <span
      className="guree-switch-inner"
    />
  </span>
  <span
    className="guree-switch-handler"
  />
  <span
    className="guree-switch-inner"
  />
</span>
`;

exports[`Switch 组件 测试 Switch 组件 defaultValue 参数 1`] = `
<span
  className="guree-switch guree-switch-checked"
  onClick={[Function]}
>
  <span
    className="guree-switch-unvisible"
  >
    <span
      className="guree-switch-handler"
    />
    <span
      className="guree-switch-inner"
    />
  </span>
  <span
    className="guree-switch-handler"
  />
  <span
    className="guree-switch-inner"
  />
</span>
`;

exports[`Switch 组件 测试 Switch 组件 disabled 参数 1`] = `
<span
  className="guree-switch guree-switch-disabled"
  onClick={[Function]}
>
  <span
    className="guree-switch-unvisible"
  >
    <span
      className="guree-switch-handler"
    />
    <span
      className="guree-switch-inner"
    />
  </span>
  <span
    className="guree-switch-handler"
  />
  <span
    className="guree-switch-inner"
  />
</span>
`;

exports[`Switch 组件 测试 Switch 组件 unCheckedChildren 参数 1`] = `
<span
  className="guree-switch"
  onClick={[Function]}
>
  <span
    className="guree-switch-unvisible"
  >
    <span
      className="guree-switch-handler"
    />
    <span
      className="guree-switch-inner"
    >
      unchecked
    </span>
  </span>
  <span
    className="guree-switch-handler"
  />
  <span
    className="guree-switch-inner"
  >
    unchecked
  </span>
</span>
`;
